packages/drivers/odsp-driver/src/epochTracker.ts

```diff
@@ -10,39 +10,111 @@ import { ThrottlingError } from "@fluidframework/driver-utils";
     "treesLatest" | "uploadSummary" | "push" | "versions";
 
+export type FetchTypeInternal = FetchType | "cache";
+
+// exported only of test purposes
+export interface IVersionedValueWithEpoch<T> {
+    value: T;
+    fluidEpoch: string,
+    version: "0.2",
+}
+
+// exported only of test purposes
+export const persistedCacheValueVersion = "0.2";
+
 /**
  * This class is a wrapper around fetch calls. It adds epoch to the request made so that the
  * server can match it with its epoch value in order to match the version.
  * It also validates the epoch value received in response of fetch calls. If the epoch does not match,
  * then it also clears all the cached entries for the given container.
  */
-export class EpochTracker {
+export class EpochTracker implements IPersistedFileCache {
     private _fluidEpoch: string | undefined;
-    private _fileEntry: IFileEntry | undefined;
+
     public readonly rateLimiter: RateLimiter;
 
     constructor(
-        private readonly persistedCache: LocalPersistentCacheAdapter,
+        protected readonly cache: IPersistedCache,
+        protected readonly fileEntry: IFileEntry,
         protected readonly logger: ITelemetryLogger,
     ) {
         // Limits the max number of concurrent requests to 24.
         this.rateLimiter = new RateLimiter(24);
     }
 
-    public set fileEntry(fileEntry: IFileEntry | undefined) {
-        assert(this._fileEntry === undefined, 0x09b /* "File Entry should be set only once" */);
-        assert(fileEntry !== undefined, 0x09c /* "Passed file entry should not be undefined" */);
-        this._fileEntry = fileEntry;
+    // public for UT purposes only!
+    public setEpoch(epoch: string, fromCache: boolean, fetchType: FetchTypeInternal) {
+        assert(this._fluidEpoch === undefined, "epoch exists");
+        this._fluidEpoch = epoch;
+
+        this.logger.sendTelemetryEvent(
+            {
+                eventName: "EpochLearnedFirstTime",
+                epoch,
+                fetchType,
+                fromCache,
+            },
+        );
+    }
+
+    public async get<T>(
+        entry: IEntry,
+    ): Promise<T | undefined> {
+        try {
+            const value: IVersionedValueWithEpoch<T> = await this.cache.get(this.fileEntryFromEntry(entry));
+            if (value === undefined || value.version !== persistedCacheValueVersion) {
+                return undefined;
+            }
+            assert(value.fluidEpoch !== undefined, "all entries have to have epoch");
+            if (this._fluidEpoch === undefined) {
+                this.setEpoch(value.fluidEpoch, true, "cache");
+            } else if (this._fluidEpoch !== value.fluidEpoch) {
+                return undefined;
+            }
+            return value.value;
+        } catch (error) {
+            this.logger.sendErrorEvent({ eventName: "cacheFetchError", type: entry.type }, error);
```

Comment:
what is type here?

Reply:
"snapshot" or "ops", i.e. what we requested. Probably does not matter that much for telemetry...